### PR TITLE
47124068_AldenShiverick_SmurfFinal

### PR DESCRIPTION
**Pull request by ashiverick**

### CODE REVIEW

**Comment by ashiverick** [comment]
Thank you for all the help

File: smurfFinalProject/smurfFinalProject/main.cpp [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    add_op                  <-  < '+' / '-' >
+    mul_op                  <-  < '*' / '/' >
+    rel_op                  <-  < '==' / '!=' / '>=' / '>' / '<=' / '<' >
+    identifier              <-  < [a-z][a-zA-Z_0-9]* >
+    call_arguments          <-  (expr (',' expr)*)?
+    param_list              <-  '(' identifier (',' identifier)* ')' / '(' ')'
+    %whitespace             <-  [ \t\r\n]*
+)";
+
+//Grammar not added yet
```
Comment (pragdave) [inline smurfFinalProject/smurfFinalProject/main.cpp:50]: Nice use of comments

File: smurfFinalProject/smurfFinalProject/main.cpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+node *bin_op(const SemanticValues &sv)                              //The binop function takes in the string passed to the parser
+{                                                                   //and breaks it into nodes and pointer nodes to create a smaller
+    node *left = sv[0].get<ParseTreeNode>().get();                  //tree within the larger AST consisting of two values as children
+    for (unsigned int i = 1; i < sv.size(); i += 2){                //and an operation node as the parent:
+        node *right = sv[i + 1].get<ParseTreeNode>().get();         //              operation
+        string op = sv[i].get<ParseTreeNode>().get()->str();        //              |       |
+        left = new binopNode(left, op, right);                      //          value       value
+    }
+    return left;
```
Comment (pragdave) [inline smurfFinalProject/smurfFinalProject/main.cpp:83]: Cool comments.  I used to do stuff like this when I was starting out.

As you continue to code, though, you'll discover that comments such as these become a burden: swap two lines in the code, and you have to adjust the comment. If you decide to split the function into two, you have to rewrite it.

Not a criticism: just a prediction :)

File: smurfFinalProject/smurfFinalProject/node.cpp [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+string codeNode::str(){                                                 //Returns the statements in strings
+    string result;
+    for(int i=0; i<statements.size(); i++){
+        result = statements[i]->str();
+    }
+    return result;
+}
+
+void codeNode::addToVect(node *nodex){                                  //Adds statements to the statements vector of nodes
```
Comment (pragdave) [inline smurfFinalProject/smurfFinalProject/node.cpp:159]: Most of your naming is great. But this one is suspect: `addToVect` describes the implementation, not the intent. Maybe `addStatement`?

File: smurfFinalProject/smurfFinalProject/main.cpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+)";
+
+//Grammar not added yet
+//expr                    <-  'if' if_expression / boolean_expression / arithmetic_expression
+//if_expression           <-  expr brace_block ( 'else' brace_block )?
+//brace_block             <-  '{' code '}'
+//function_definition     <-  param_list brace_block
+//comment                <-  '#' r'.*'
+
+class visitor;                                                      //The visitor pattern goes to each node to evaluate
```
Comment (pragdave) [inline smurfFinalProject/smurfFinalProject/main.cpp:57]: Visitor is an abstract base class: you don't need to instantiate it...

File: smurfFinalProject/smurfFinalProject/main.cpp [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    };
+    
+    parser["identifier"] = [](const SemanticValues &sv) {
+        return ParseTreeNode(new variableNode(sv.token()));
+    };
+    
+    parser["expr"] = [](const SemanticValues &sv) {
+        node *n = bin_op(sv);
+        return ParseTreeNode(n);
+    };
```
Comment (pragdave) [inline smurfFinalProject/smurfFinalProject/main.cpp:139]: An expr isn't a binop. This still works: can you work out why?

But this clause, and the let_stmt clause above, actually do nothing; your parser works fine if I remove them both.

# -2

**Review by pragdave — COMMENTED** [review]
Deducting 2 because you're not showing me the full AST.

**Comment by pragdave** [comment]
Grading this code is like judging an old friend :)

I have been really impressed by your commitment to this project, and by the way you stuck with it through all the frustrations and problems, some of which were not of your own making. People tend to judge developers purely on technical skills, but in reality that's just a part of the equation. When working in the real world, on real teams, attitude and courage are just as valuable (if not more so). You've shown both over the last month. You've also exhibited a willingness to ask for help, which seems to be rare but also incredibly important.

For all of these reasons:  

# +5

**Pull request closed without merging**